examples/cpp/route_guide/route_guide_callback_server.cc

```diff
@@ -228,25 +230,21 @@ class RouteGuideImpl final : public RouteGuide::CallbackService {
 
      private:
       void NextWrite() {
-        while (notes_iterator_ != received_notes_->end()) {
-          const RouteNote& n = *notes_iterator_;
+        if (notes_iterator_ != to_send_notes_.end()) {
+          StartWrite(&*notes_iterator_);
           notes_iterator_++;
-          if (n.location().latitude() == note_.location().latitude() &&
-              n.location().longitude() == note_.location().longitude()) {
-            StartWrite(&n);
-            return;
-          }
+        } else {
+          mu_->lock();
+          received_notes_->push_back(note_);
+          mu_->unlock();
+          StartRead(&note_);
         }
-        // Didn't write anything, so all done with this note
-        received_notes_->push_back(note_);
-        mu_->unlock();
-        StartRead(&note_);
       }
       RouteNote note_;
       std::mutex* mu_;
       std::vector<RouteNote>* received_notes_;
```

Comment:
In general these feels like a back way to build this project up, shouldn't it be more like a map or something so that you're not going through the whole thing  every time since you only have to send the notes associated with a particular point?

Reply:
That's... true, I guess. I didn't necessarily want to rewrite the whole thing tho, only fix the incorrect mutex usage :)